### ISSUE TITLE
feat: lightweight, native ann-vector database for long-term memory

### ISSUE THREAD

**Issue by jon-chuang** [rustformers/llama-rs#133]
I think the next step in the project is lightweight ANN (approx. k-nearest neighbours search) vector database. Applications:
1. Document store over local documents: code bases, journals, articles. Input: directory with text files.
2. Chrome-assistant: A memory over your currently and recently opened tabs
3. Mobile: similar to local

Problem definition:
1. We are ok trading off a bit of performance to have something that has minimal surface area
2. We want the database to be persistent - it should persist the index after generation. The model is very similar to ggml, we should generate the artifact once, or periodically, and then load the artifact (index) into memory. We can use mmap trick for better performance just like ggmt.

Options:
1. connect to existing vector database (qdrant, milvus, pinecone). But these are heavy deps, and also have many features designed that are more about scaling out (cloud native). We like transparency and owning the artifacts involved. We are willing to tradeoff a bit of perf and or implementation complexity for this.
2. Compile faiss as an optional dependency. Still a pretty huge dependency.
3. Something rust-native: e.g. https://github.com/hora-search/hora. Not clear how well maintained.

We need to develop a persistent format for index for 2. and 3.

**Comment by hhamud** [comment]
interesting, like a rust specific version of [this](https://github.com/chroma-core/chroma)?

**Comment by jon-chuang** [comment]
> re-use previous prompts or their outputs

The problem with a hash table or KV store is that natural language queries are rarely exactly the same, especially if you are not averaging over the human population but just running local.

**Comment by philpax** [comment]
I think this is out of scope for this repository specifically. I could see a batteries-included implementation being built atop `llama-rs`, but it's unlikely to feature an implementation of a vector database itself because our focus is specifically on robust, fast inference of LLMs.

**Comment by jon-chuang** [comment]
> our focus is specifically on robust, fast inference of LLMs.

Yes, but I think the broader focus is "low-resource, low-dependency embedded LLM toolchain".

I can definitely see the sliced out k-ANN code existing in a separate repo (perhaps under this org) and compiled in as an optional dependency to the cli (on crates.io it would be `cargo install llama-rs --features "knowledge-base"`)

**Comment by hhamud** [comment]
> "low-resource, low-dependency embedded LLM toolchain".

Literally what I was thinking of yesterday

**Comment by jon-chuang** [comment]
I've made an issue here to sound out the idea: https://github.com/ggerganov/llama.cpp/issues/930

**Comment by philpax** [comment]
> I can definitely see the sliced out k-ANN code existing in a separate repo (perhaps under this org) and compiled in as an optional dependency to llama-rs and available in the cli (on crates.io it would be `cargo install llama-rs --features "knowledge-base"`)

Sure, but I don't see why it would have to be part of `llama-rs` specifically. The CLI is really just a demo application for the library; it doesn't aspire to higher functionality than that.

I'm not opposed to having this kind of functionality - having a full-stack solution for using a LLM to do knowledge base inference would be great - but I think it's a hard sell to make it part of this crate specifically. By analogy, we're like `hyper`, not `reqwest` - we're not trying to solve all the problems, just the core problem that enables other people to solve their problems.

**Comment by jon-chuang** [comment]
> but I think it's a hard sell to make it part of this crate specifically.

I'm in agreement here. But do you think that rustformers more generally could be expanded to the scope of "low-resource, low-dependency embedded LLM toolchain" and host the `llama-rs-toolchain` which builds such features around llama-rs?
